### PR TITLE
Strengthen Request validation

### PR DESCRIPTION
**Pull request by geoknee**
Has the side effect that concluding now works :-).

### CODE REVIEW

File: packages/client-api-schema/src/__tests__/good_sample_requests.ts [hunk 1]
```diff
@@ -95,8 +95,14 @@ export const createChannel = {
       {
         token: '0x0',
         allocationItems: [
-          {destination: '0x63e3fb11830c01ac7c9c64091c14bb6cbaac9ac7', amount: '0x06f05b59d3b20000'},
-          {destination: '0x63e3fb11830c01ac7c9c64091c14bb6cbaac9ac7', amount: '0x06f05b59d3b20000'}
+          {
+            destination: '0x63e3fb11830c01ac7c9c64091c14bb6cbaac9ac7',
+            amount: '0x00000000000000000000000000000000000000000000000006f05b59d3b20000'
```
Comment (andrewgordstewart) [inline packages/client-api-schema/src/__tests__/good_sample_requests.ts]: Why does this need to be padded?

File: packages/client-api-schema/src/generated-schema.json [hunk 2]
```diff
@@ -1401,6 +1406,11 @@
       ],
       "type": "object"
     },
+    "Uint256": {
+      "description": "Uint256",
+      "pattern": "^0x([a-fA-F0-9]{64})$",
```
Comment (andrewgordstewart) [inline packages/client-api-schema/src/generated-schema.json]: I don't think this needs to be 32 bytes. 
```suggestion
      "pattern": "^0x([a-fA-F0-9]{0,64})$",
```

**Comment by geoknee** [comment]
Painful: `ajv` errors are hardly human readable. I had tried using https://github.com/atlassian/better-ajv-errors which is much better, but unfortunately not support in the browser at present. 

Horrible: so many `hexZeroPad` everywhere :-(

File: packages/client-api-schema/src/types.ts [hunk 3]
```diff
@@ -16,6 +16,18 @@ export type ChannelId = string;
...
 
+/**
+ * Bytes32
+ * @pattern  ^0x([a-fA-F0-9]{64})$
+ */
+export type Bytes32 = string;
+
+/**
+ * Uint256
+ * @pattern  ^0x([a-fA-F0-9]{64})$
```
Comment (andrewgordstewart) [inline packages/client-api-schema/src/types.ts]:
```suggestion
 * @pattern  ^0x([a-fA-F0-9]{0,64})$
```

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 4]
```diff
@@ -318,8 +320,14 @@ const convertToChannelState = (channelResult: ChannelResult): ChannelState => {
...
-    beneficiaryBalance: bigNumberify(allocations[0].allocationItems[0].amount).toString(),
-    payerBalance: bigNumberify(allocations[0].allocationItems[1].amount).toString()
+    beneficiaryBalance: hexZeroPad(
+      bigNumberify(allocations[0].allocationItems[0].amount).toHexString(),
+      32
+    ),
+    payerBalance: hexZeroPad(
+      bigNumberify(allocations[0].allocationItems[1].amount).toHexString(),
+      32
+    )
```
Comment (andrewgordstewart) [inline packages/web3torrent/src/clients/payment-channel-client.ts]: Unnecessary if the validation schema is changed as per above.

File: packages/xstate-wallet/src/workflows/conclude-channel.ts [hunk 5]
```diff
@@ -19,6 +19,7 @@ const finalState = (store: Store) => async (context: Init): Promise<SupportState
 
   // If we've received a new final state that matches our outcome we support that
   if (outcomesEqual(latestSupportedByMe.outcome, latestFinalState?.outcome)) {
+    // outcomesEqual function will return false if there is inconsistent padding of hex strings
```
Comment (andrewgordstewart) [inline packages/xstate-wallet/src/workflows/conclude-channel.ts]: What is this comment for? It's my understanding that `outcomesEqual` should _not_ return `false` in this case.

Comment (geoknee) [reply packages/xstate-wallet/src/workflows/conclude-channel.ts]: The comment is telling you that the function does not do what you think it should.

Comment (andrewgordstewart) [reply packages/xstate-wallet/src/workflows/conclude-channel.ts]: Shouldn't it be on `outcomesEqual` then?

File: packages/client-api-schema/src/types.ts (hunk 3 above)
Comment (geoknee) [reply packages/client-api-schema/src/types.ts]: I copied this from the wire-format validator in the xstate wallet.

File: packages/client-api-schema/src/generated-schema.json (hunk 2 above)
Comment (geoknee) [reply packages/client-api-schema/src/generated-schema.json]: See below comment (this file autogenerated anyway)

File: packages/client-api-schema/src/__tests__/good_sample_requests.ts (hunk 1 above)
Comment (geoknee) [reply packages/client-api-schema/src/__tests__/good_sample_requests.ts]: In order to pass the schema after the latter changed (merit of which we are discussing in other threads on this page).

File: packages/xstate-wallet/src/workflows/conclude-channel.ts (hunk 5 above)
Comment (geoknee) [reply packages/xstate-wallet/src/workflows/conclude-channel.ts]: I propose that we do as you suggest and make this function behave independently of the formatting of the outcomes, say by using `bigNumberify`.

Comment (geoknee) [reply packages/xstate-wallet/src/workflows/conclude-channel.ts]: (And then remove the comment)

File: packages/client-api-schema/src/types.ts (hunk 3 above)
Comment (geoknee) [reply packages/client-api-schema/src/types.ts]: If you have fewer than 64 nibbles, then you have fewer than 256 bits.

And if you have something other than a multiple of 8 nibbles, then you don't have a valid solidity data type at all.

 Of course most things would likely still work regardless, but if we are going to call something a Uint256, it seems to me that it should have 256 bits.

File: packages/xstate-wallet/src/workflows/conclude-channel.ts (hunk 5 above)
Comment (geoknee) [reply packages/xstate-wallet/src/workflows/conclude-channel.ts:23]: #1342

**Pull request merged**